pin the PEX_PYTHON{,_PATH} running the pytest pex to avoid using incompatible pytest requirements

cosmicexplorer opened the pull request:
### Problem

`PytestPrep` will generate a pex merging python sources and 3rdparty requirements. This pex will have all of the unique python interpreter constraints from all targets applied to the pex before being built. When executing that pex, it appears that the wrong python interpreter may be selected when the interpreter constraint set is intersecting -- see the added test `test_succeeds_for_intersecting_unique_constraints()`. This ends up looking like:
```
Failed to execute PEX file, missing macosx_10_14_x86_64-cp-27-cp27m compatible dependencies for:
more-itertools
pytest
funcsigs
pytest-cov
coverage
```

### Solution

- Force the selected interpreter used to create the pytest pex to be used when running it by setting `PEX_PYTHON` and `PEX_PYTHON_PATH`.
- Add testing for the case of targets with intersecting interpreter constraints (such as `['CPython>=2.7,<4', 'CPython>=3.6']`).

### Result

Running python tests targets with overlapping interpreter constraints should work!

cosmicexplorer commented:
> Thanks! would be nice to have some tests if they are not too expensive.

I think that's reasonable, especially given that some tests are failing in the current CI run. Investigating those failures might be able to produce a new test that checks for the corrected behavior.

jsirois commented on src/python/pants/backend/python/tasks/pytest_run.py:
```diff
@@ -514,8 +516,14 @@ def _do_run_tests_with_args(self, pex, args):
+      # Ensure the pex is invoked with the interpreter we selected beforehand to be compatible with
+      # all the relevant targets.
+      interpreter = self.context.products.get_data(PythonInterpreter)
+      env['PEX_PYTHON'] = interpreter.binary
```
Te pexrc issue seems out of hand and this seems like a bandaid.

Is this more appropriate?
```
$ pex --help-variables
...
PEX_IGNORE_RCFILES: Boolean

    Explicitly disable the reading/parsing of pexrc files (~/.pexrc).
    Default: false.
...
```
It seems like we should be setting this when invoking any pex-based tool in pants, not just here.

blorente replied:
I thought that flag was on its way out: https://github.com/pantsbuild/pex/pull/673, but if it's okay to use it, then it looks like the thing we want to use.

jsirois replied:
Yes, the flag, not the option. Flags for pex like --rcfile are only used at build time. options like PEX_IGNORE_RCFILES at runtime. For some options, specifying them at build time or run time both make sense. For this one, only run time makes sense.

blorente replied:
Thanks for the clarification!

cosmicexplorer replied:
The flag `PEX_IGNORE_RCFILES` has not appeared to have an effect on the case I was looking at (and the case I was looking at didn't appear to be due to the use of a pexrc!). I was able to isolate a repro and convert it into the new `test_succeeds_for_intersecting_unique_constraints()`, which this PR now fixes.

cosmicexplorer commented:
Refactored this after a while experimenting with the failure we were seeing internally, the result manages to fix the issue without introducing further breakage. It appears the issue was probably not due to a pexrc, and was instead a more general issue, which is now fixed.

This should be ready for review again. Commits are reviewable independently.

cosmicexplorer commented:
The CI failure is legit, I'm fixing that now.

cosmicexplorer commented:
Ok, so I just increased the `--timeout-default` argument for the one failing PytestRun integration test, since the previous failure actually did seem to be flakiness (and it had happened twice in a row). It's suspicious that this change would seem to cause that specific flakiness, and we will be doing performance testing of this change internally to avoid introducing a regression.

cosmicexplorer commented:
Ok, the two pex issues you linked seem to match this failure perfectly and the larger context is clear.

> NB#2: All the work in this PR is regrettably whac-a-mole. It seems to me it would be a good idea to file an issue to absract out a ~PexExecutor that all of pants can use to hermetically execute a pex using a selected interpreter just like the bits of changed code in this PR eventually will.

Well, in this one I could at least move the functionality into a separate `PexExecutor` class so other tasks can pick it up when they need it.

> NB: That change does not deal with the possibility of pexrc.

I believe the pexrc config is not affecting the failure here, or at least, setting `PEX_IGNORE_RCFILES` and using the python interpreter in the argv does not work in the way that setting `PEX_PYTHON_PATH` and `PEX_PYTHON` does. On my laptop the changes in 5a461db (setting just `PEX_PYTHON_PATH` and explicitly selecting the interpter) caused the pytest invocation to fail before tests were run:
```
============== test session starts ===============
platform darwin -- Python 3.6.4, pytest-3.6.4, py-1.8.0, pluggy-0.7.1 -- /Users/dmcclanahan/tools/pants/build-support/pants_dev_deps.py36.venv/bin/python3
cachedir: .pytest_cache
rootdir: /Users/dmcclanahan/tools/pants, inifile: /Users/dmcclanahan/tools/pants/.pants.d/test/pytest-prep/CPython-3.6.4/667aa724ef06c0a4cda5a996a39591b15fbcde5a/pytest.ini
plugins: cov-2.4.0, timeout-1.2.1
collecting ...
- generated xml file: /Users/dmcclanahan/tools/pants/.pants.d/test/pytest/tests.python.pants_test.backend.python.tasks.pytest_run/junitxml/TEST-tests.python.pants_test.backend.python.tasks.pytest_run.xml -
========== no tests ran in 0.04 seconds ==========
ERROR: not found: /Users/dmcclanahan/tools/pants/build-support/pants_dev_deps.py36.venv/bin/python3
(no name '/Users/dmcclanahan/tools/pants/build-support/pants_dev_deps.py36.venv/bin/python3' in any of [])


                   tests/python/pants_test/backend/python/tasks:pytest_run    .....   NOT RUN
```

> If you have pexrc files that specify both PEX_PYTHON and PEX_PYTHON_PATH

I do. Unfortunately setting `PEXRC_IGNORE_FILES=1` in 5a461db failed to run tests, while explicitly setting both `PEX_PYTHON_PATH` and `PEX_PYTHON` successfully ran tests.

I think regardless of the eventual fix, generalizing it into a `PexExecutor` to pin the interpreter seems like a good idea.

jsirois commented:
> Well, in this one I could at least move the functionality into a separate PexExecutor class so other tasks can pick it up when they need it.

I encourage you to fight your nature here and maintain scope. It's not just for you, it's for reviewers. It's nice for PRs to close out ~promptly as the norm. I'd just file an issue.

cosmicexplorer commented:
@jsirois I've broken out abstracting hermetic pex execution into #7640, and while I haven't determined why using `PEX_PYTHON_PATH=... /path/to/python /path/to/pex.pex args...` attempting in 9a68c3c is failing, the rationale for instead having to set both `PEX_PYTHON` and `PEX_PYTHON_PATH` is described in the docstring for `_constrain_pytest_interpreter_search_path()`. The added test demonstrates a case which is currently broken that is now fixed, and this diff fixes the same breakage we were seeing internally. Is it reasonable to try to merge this PR now, or am I missing anything?

stuhood commented on src/python/pants/backend/python/tasks/python_execution_task_base.py:
```diff
@@ -116,8 +116,17 @@ def create_pex(self, pex_info=None):
+        }
+        self.context.log.debug('unique_constraints:\n{}'.format(unique_constraints))
+
+        if pin_selected_interpreter:
```
This flag seems to disable the `unique_constraints` codepath entirely. should those be computed under that branch of the if instead?

It seems like rather than "pinning", this is more like: "use_global_python_interpreter" or something, because what it pins to is specifically the `PythonInterpreter` product. It should definitely have a mention in the docstring.

cosmicexplorer replied:
I've removed the extra argument to `create_pex()` in favor of removing the `unique_constraints` code path entirely, as every consumer of this method is always going to depend on requirements pexes resolved specifically for the single global selected interpreter. I've also added a docstring clarifying all of this.

cosmicexplorer replied:
As noted in https://github.com/pantsbuild/pants/pull/7563#issuecomment-488205889, I've added back the `pin_selected_interpreter` argument, but added a descriptive docstring. I'll be following up with a change that removes this.

jsirois commented on src/python/pants/backend/python/tasks/pytest_run.py:
```diff
@@ -482,6 +483,21 @@ def _test_runner(self, workdirs, test_targets, sources_map):
+    the interpreter search path at pex runtime ensures that any resolved requirements will be
+    compatible with the interpreter being used to invoke the merged pytest pex.
+    """
+    pytest_prep_binary_product = self.context.products.get_data(PytestPrep.PytestBinary)
```
You're using java words like I would tend to - this is surprising. Just `pytest_binary` seems fine since no line below care this is a product - just this line here where the RHS makes that clear anyway.

cosmicexplorer replied:
I'll make this fix, and will keep this in mind. I've been trying to avoid abbreviations in variable names, but writing out irrelevant information as you've described is also not ideal.

jsirois commented on tests/python/pants_test/task_test_base.py:
```diff
@@ -380,8 +382,12 @@ def invoke_tasks(self, target_closure=None, **context_kwargs):
+      for tsk in all_task_instances:
+        tsk.execute()
+    except Exception as e:
+      setattr(e, '_context', context)
```
This is super hacky. Consider circling back in a follow up before reliance on this gets out of control and either
1. splitting the method into two parts, one that returns context, run_before_synthesized_task_types, run_after_synthesized_task_types and one that consumes these 3 and actually invokes tasks
2. Always return a self.TaskInvocationResult but teach that type to say success/error or make it an enum or what have you.

cosmicexplorer replied:
I would normally do (1), but since we already have the special `self.TaskInvocationResult` class we might as well use it to actually represent all types of results from invoking tasks, so will likely do (2) in a followup. Thank you for outlining how to effectively address this!

cosmicexplorer commented:
I'm going to revert the changes to `create_pex()` which are causing the one CI failure and break it out into a followup PR.

jsirois commented on src/python/pants/backend/python/tasks/python_execution_task_base.py:
```diff
@@ -127,7 +134,16 @@ def create_pex(self, pex_info=None):
+        else:
+          constraints = {
+            constraint for rt in relevant_targets if is_python_target(rt)
+          for constraint in PythonSetup.global_instance().compatibility_or_constraints(rt)
```
I'm not sure if this works with python ws rules - probably does - but worth fixing the indent anyhow. I suspect you didn't intend this.

cosmicexplorer replied:
That's correct, fixing! Thanks!
